### PR TITLE
Add initial implementation

### PR DESCRIPTION
**Pull request by damdo**
This adds the initial implementation for the gokrazy selfupdate package.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+// readConfigFile reads configuration files from /perm /etc or / and returns
+// trimmed content as string.
+//
+// TODO: de-duplicate this with gokrazy.go into a gokrazy/internal package
```
Comment (stapelberg) [inline config.go]: Can you update the comment: doesn’t need to be gokrazy/internal, could even be useful for public packages

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,15 @@
+module github.com/gokrazy/selfupdate
+
+go 1.19
```
Comment (stapelberg) [inline go.mod]: go 1.20?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	"syscall"
+	"time"
+
+	"github.com/gokrazy/gokrazy"
+)
+
+var (
+	updateEndpoint string
+	checkInterval  string
+)
```
Comment (stapelberg) [inline main.go]: Can you move these into main()?

File: main.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	log.Print("gokrazy's selfupdate service starting up..")
+
+	flag.StringVar(&updateEndpoint, "update_endpoint", "", "the HTTP/S endpoint of the update service (required)")
+	flag.StringVar(&checkInterval, "check_interval", "1h", "the time duration interval between checks to the update service. default: 1h")
+
+	flag.Parse()
+
```
Comment (stapelberg) [inline main.go:38]: Can you move everything below into a `func selfupdate() error { …}`?

File: main.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		log.Fatalf("could read neither /perm/gokr-pw.txt, nor /etc/gokr-pw.txt, nor /gokr-pw.txt: %s", err.Error())
+	}
+
+	httpPort, err := readConfigFile("http-port.txt")
+	if err != nil {
+		log.Fatalf("could read neither /perm/http-port.txt, nor /etc/http-port.txt, nor /http-port.txt: %s", err.Error())
+	}
+
+	log.Printf("entering update checking loop with interval: %s", interval.String())
+	ticker := time.NewTicker(interval)
```
Comment (stapelberg) [inline main.go]: Can you add jitter in the range of a few seconds to minutes? https://github.com/gokrazy/gokrazy/blob/9cdaabefeec8cd194f0a14409f0b85f1c6217540/cmd/heartbeat/heartbeat.go#L133

File: main.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+				if err := selfupdate(ctx, response, httpPassword, httpPort); err != nil {
+					log.Fatal(err)
+				}
+
+				// The update is now correctly written to the disk partitions
+				// and the reboot is in progress
+				// sleep until the context chan is closed, then exit cleanly.
+				<-ctx.Done()
+				os.Exit(0)
+			}
```
Comment (stapelberg) [inline main.go]: Can you refactor this into a standalone function?

File: selfupdate.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, fmt.Errorf("error json encoding request: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", updateEndpoint, bytes.NewBuffer(reqBody))
+	if err != nil {
+		return nil, fmt.Errorf("error creating http request: %w", err)
+	}
+	req.Header.Set("Content-Type", "text/json")
```
Comment (stapelberg) [inline selfupdate.go]: We should set both headers: Accept and Content-Type. 

The MIME type is application/json, not text/json

File: selfupdate.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+	var rootReader, bootReader, mbrReader io.ReadCloser
+	var zipReader *zip.ReadCloser
+	var err error
+
+	switch response.RegistryType {
+	case "http", "localdisk":
+		zipReader, mbrReader, bootReader, rootReader, err = httpHandler(response)
+		if err != nil {
+			return fmt.Errorf("error fetching '%s' update from link '%s': %w", response.RegistryType, response.Link, err)
```
Comment (stapelberg) [inline selfupdate.go]: here and elsewhere: Use %q instead of '%s'

File: selfupdate.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		return fmt.Errorf("unrecognized registry type '%s'", response.RegistryType)
+	}
+
+	uri := fmt.Sprintf("http://gokrazy:%s@127.0.0.1:%s/", httpPassword, httpPort)
+
+	log.Print("checking target partuuid support")
+
+	target, err := updater.NewTarget(uri, http.DefaultClient)
+	if err != nil {
+		log.Fatalf("checking target partuuid support: %v", err)
```
Comment (stapelberg) [inline selfupdate.go]: Here and below: use return fmt.Errorf instead of log.Fatalf directly

File: update_handlers.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+)
+
+const (
+	mbrPartitionName  = "mbr.img"
+	bootPartitionName = "boot.img"
+	rootPartitionName = "root.img"
+)
+
+// httpHandler handles a http update link.
+func httpHandler(response *updateResponse) (*zip.ReadCloser, io.ReadCloser, io.ReadCloser, io.ReadCloser, error) {
```
Comment (stapelberg) [inline update_handlers.go]: httpFetcher?

File: update_handlers.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	// when necessary.
+	link, err := ensureAbsoluteHTTPLink(updateEndpoint, response.Link)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("error ensuring absolute HTTP link '%s' + '%s': %w",
+			updateEndpoint, response.Link, err)
+	}
+
+	log.Printf("downloading update file from registry '%s' with url: %s", response.RegistryType, link)
+
+	filePath := "/tmp/drive.zip"
```
Comment (stapelberg) [inline update_handlers.go]: Can you add a TODO for the location on disk (or on RAM)?

File: update_handlers.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	// The link may be a relative url if the server's backend registry is its local disk.
+	// Ensure we have an absolute url by adding the base (updateEndpoint) url
+	// when necessary.
+	link, err := ensureAbsoluteHTTPLink(updateEndpoint, response.Link)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("error ensuring absolute HTTP link '%s' + '%s': %w",
+			updateEndpoint, response.Link, err)
+	}
+
+	log.Printf("downloading update file from registry '%s' with url: %s", response.RegistryType, link)
```
Comment (stapelberg) [inline update_handlers.go]: Should we check for available disk space? Maybe leave a TODO

File: update_handlers.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+// httpDownloadFile downloads a file from a url to a filepath on disk.
+// It's efficient because it will write as it downloads
+// and not load the whole file into memory.
+func httpDownloadFile(filepath string, url string) error {
+	// Get the data
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
```
Comment (stapelberg) [inline update_handlers.go:95]: Check the resp.StatusCode to be HTTP 200 OK

File: update_handlers.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	// Create the file
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	// Write the body to file
+	_, err = io.Copy(out, resp.Body)
+	return err
```
Comment (stapelberg) [inline update_handlers.go]: Check the error that file.Close() returns:

if err := out.Close(); err != nil {

File: update_handlers.go (hunk 10 above)
Comment (stapelberg) [inline update_handlers.go]: Maybe return a struct type instead of 4 different return values

**Review by stapelberg — CHANGES_REQUESTED** [review]

File: main.go [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+	httpPort, err := readConfigFile("http-port.txt")
+	if err != nil {
+		return fmt.Errorf("could read neither /perm/http-port.txt, nor /etc/http-port.txt, nor /http-port.txt: %s", err.Error())
+	}
+
+	if o.skipWaiting {
+		log.Print("skipping waiting, performing an immediate updateProcess")
+		if err := updateProcess(ctx, &updateRequest{MachineID: machineID, SBOMHash: sbomHash}, o.updateEndpoint, sbomHash, o.destinationDir, httpPassword, httpPort); err != nil {
+			// If the updateProcess we exit with an error
```
Comment (stapelberg) [inline main.go]: missing “fails”

File: main.go [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+			// so that gokrazy supervisor will restart the process.
+			return fmt.Errorf("error performing updateProcess: %v", err)
+		}
+
+		// If the updateProcess doesn't error
+		// we happily terminate the process.
+		os.Exit(0)
```
Comment (stapelberg) [inline main.go]: return nil?

File: main.go [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	log.Print("entering update checking loop")
+	ticker := time.NewTicker(frequency)
+
+	for loop := true; loop; {
+		select {
+		case <-ctx.Done():
+			log.Print("stopping update checking")
+			ticker.Stop()
```
Comment (stapelberg) [inline main.go]: return nil?

File: main.go [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+
+func updateProcess(ctx context.Context, upReq *updateRequest, updateEndpoint, sbomHash, destinationDir, httpPassword, httpPort string) error {
+	response, err := checkForUpdates(ctx, updateEndpoint, upReq)
+	if err != nil {
+		return fmt.Errorf("unable to check for updates: %w", err)
+	}
+
+	// Check if we should update by comparing the update response SBOMHash with
+	// the current installation SBOMHash.
+	if shouldUpdate(response, sbomHash) {
```
Comment (stapelberg) [inline main.go]: inverse conditional:
if !shouldUpdate(…) {
return nil
}
// remaining logic

File: selfupdate.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error making http request: %w", err)
+	}
+	if resp.StatusCode != http.StatusOK {
```
Comment (stapelberg) [inline selfupdate.go:46]: defer resp.Body.Close() above line 44

File: selfupdate.go [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+	case "http", "localdisk":
+		readClosers, err = httpFetcher(response, updateEndpoint, destinationDir)
+		if err != nil {
+			return fmt.Errorf("error fetching %q update from link %q: %w", response.RegistryType, response.Link, err)
+		}
+	default:
+		return fmt.Errorf("unrecognized registry type %q", response.RegistryType)
+	}
+
+	uri := fmt.Sprintf("http://gokrazy:%s@127.0.0.1:%s/", httpPassword, httpPort)
```
Comment (stapelberg) [inline selfupdate.go]: Use “localhost”

File: main.go [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	log.Print("gokrazy's selfupdate service starting up..")
+
+	var o opts
+
+	flag.StringVar(&o.updateEndpoint, "update_endpoint", "", "the HTTP/S endpoint of the update service (required)")
+	flag.StringVar(&o.checkFrequency, "check_frequency", "1h", "the time frequency for checks to the update service. default: 1h")
+	flag.StringVar(&o.destinationDir, "destination_dir", "/tmp", "the destination directory for the fetched update file. default: /tmp")
```
Comment (stapelberg) [inline main.go]: Make the default /tmp/selfupdate

File: update_handlers.go [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+	availableDiskSpace, err := diskSpaceAvailable(destinationDir)
+	if err != nil {
+		return rcs{}, fmt.Errorf("error checking available disk space for %q", destinationDir)
+	}
+
+	filePath := filepath.Join(destinationDir, "drive.gaf")
+
+	fileSize, err := httpFileSize(link)
+	if err != nil {
+		return rcs{},
```
Comment (stapelberg) [inline update_handlers.go]: remove linebreak

File: update_handlers.go [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+
+// httpDownloadFile downloads a file from a url to a filepath on disk.
+// It's efficient because it will write as it downloads
+// and not load the whole file into memory.
+func httpDownloadFile(filepath string, url string) error {
+	// Get the data.
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
```
Comment (stapelberg) [inline update_handlers.go:94]: resp.StatusCode check?

File: update_handlers.go [hunk 24]
```diff
@@ -0,0 +1,166 @@
...
+	// when necessary.
+	link, err := ensureAbsoluteHTTPLink(updateEndpoint, response.Link)
+	if err != nil {
+		return rcs{}, fmt.Errorf("error ensuring absolute HTTP link %q + %q: %w",
+			updateEndpoint, response.Link, err)
+	}
+
+	log.Printf("downloading update file from registry %q with url: %s", response.RegistryType, link)
+
+	availableDiskSpace, err := diskSpaceAvailable(destinationDir)
```
Comment (stapelberg) [inline update_handlers.go]: Move this into httpDownloadFile to save 1 HTTP request

File: update_handlers.go [hunk 25]
```diff
@@ -0,0 +1,166 @@
...
+		return err
+	}
+	defer out.Close()
+
+	// Write the body to file.
+	if _, err := io.Copy(out, resp.Body); err != nil {
+		return fmt.Errorf("error writing file %q to disk: %w", filepath, err)
+	}
+
+	if err := resp.Body.Close(); err != nil {
```
Comment (stapelberg) [inline update_handlers.go]: This should be out.Close(), not resp.Body.Close()

File: update_handlers.go (hunk 23 above)
Comment (stapelberg) [inline update_handlers.go:94]: defer resp.Body.Close

**Review by stapelberg — CHANGES_REQUESTED** [review]

File: update_handlers.go [hunk 26]
```diff
@@ -37,28 +37,14 @@ func httpFetcher(response *updateResponse, updateEndpoint, destinationDir string
...
 
-	if availableDiskSpace < fileSize {
-		return rcs{}, fmt.Errorf("error: refused downloading %q to %q,"+
-			"not enough space left on the device. required: %d available %d",
-			link, filePath, fileSize, availableDiskSpace)
-	}
+	log.Printf("downloading update file from registry %q with url: %s", response.RegistryType, link)
 
-	if err := httpDownloadFile(filePath, link); err != nil {
+	filePath := filepath.Join(destinationDir, "drive.gaf")
```
Comment (stapelberg) [inline update_handlers.go]: disk.gaf

File: main.go [hunk 27]
```diff
@@ -88,9 +88,7 @@ func logic(ctx context.Context, o opts) error {
 		select {
 		case <-ctx.Done():
 			log.Print("stopping update checking")
-			ticker.Stop()
-			loop = false
```
Comment (stapelberg) [inline main.go]: The loop variable can now be removed from the loop in line 87, too

File: update_handlers.go [hunk 28]
```diff
@@ -96,27 +82,47 @@ func httpFetcher(response *updateResponse, updateEndpoint, destinationDir string
 // httpDownloadFile downloads a file from a url to a filepath on disk.
 // It's efficient because it will write as it downloads
 // and not load the whole file into memory.
-func httpDownloadFile(filepath string, url string) error {
-	// Get the data.
+func httpDownloadFile(destinationDir, filePath string, url string) error {
+
```
Comment (stapelberg) [inline update_handlers.go]: remove blank line

File: update_handlers.go [hunk 29]
```diff
@@ -96,27 +82,47 @@ func httpFetcher(response *updateResponse, updateEndpoint, destinationDir string
...
+	}
+
 	resp, err := http.Get(url)
 	if err != nil {
 		return err
 	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("unexpected status code received: %d", resp.StatusCode)
```
Comment (stapelberg) [inline update_handlers.go]: return fmt.Errorf("unexpected HTTP status code: %v", resp.Status)

**Review by stapelberg — CHANGES_REQUESTED** [review]

**Comment by damdo** [comment]
Feedback addressed and commits now squashed.

**Comment by damdo** [comment]
Once this merges we need to change the default branch from init to main in the GH preferences.

**Review by stapelberg — APPROVED** [review]

**Pull request merged**